### PR TITLE
Issue 263/make tabbed layout configurable

### LINKED ISSUE #263: Make organize tabs layout configurable
The tabs needed are different for different pages -- currently the tab values are hardcoded into the component, make it configurable

### LINKED ISSUE #296: using OrganizeTabbedLayout on all campaigns page results in a bad call to useQuery
When using the tabbed layout on all campaigns pages, useQuery tries to fetch single campaign data with `undefined` as the campId, resulting in a console error (and a needless call). 

Putting the hook inside a conditional raises other warnings related to react. 

There are probably multiple solutions to this

### PR DESCRIPTION
**Pull request by nehahirve**
Resolves #263 and Resolves #296

Renaming: OrganizeLayout and OrganizeTabbedLayout are now DefaultLayout and TabbedLayout and lie in `layout/organize`

### CODE REVIEW

File: src/components/layout/organize/TabbedLayout.tsx [hunk 1]
```diff
@@ -22,41 +20,38 @@ const useStyles = makeStyles((theme) => ({
     },
 }));
 
-interface OrganizeTabbedLayoutProps {
+interface TabbedLayoutProps {
     fixedHeight?: boolean;
+    title?: string;
+    subtitle?: string;
+    baseHref: string;
+    tabs: { defaultTab?:boolean; href: string; label: string}[];
```
Comment (richardolsson) [inline src/components/layout/organize/TabbedLayout.tsx:28]: The downside with the `defaultTab` boolean is that it could be true for multiple tabs, and behavior is then not clear (although I can see by reading the code what would happen). 

A better design IMO is having some sort of unique key for the tabs (e.g. the `label` or the `href` can be considered that key) and then a separate `defaultTab` property where the value should be the key of that tab.

File: src/components/layout/organize/TabbedLayout.tsx [hunk 2]
```diff
@@ -22,41 +20,38 @@ const useStyles = makeStyles((theme) => ({
...
 
-    const path =  router.pathname.split('/');
-    let currentTab = path.pop();
-
-    if (currentTab === '[campId]' || currentTab === 'campaigns') {
-        currentTab = 'summary';
+    if (router.asPath === baseHref) {
+        currentTab = tabs.find(tab => tab.defaultTab)?.label;
+        if (!currentTab) {
+            console.warn('No default tab selected.');
```
Comment (richardolsson) [inline src/components/layout/organize/TabbedLayout.tsx:41]: I'm not a big fan of these. If not having a default tab is a bug, it should be louder than a warning to developers, and ideally there should be no console output to non-developers.

I would suggest instead throwing an error if `NODE_ENV` is `development`, and else do nothing.

Even better of course would be some sort of compile-time type safety, but I'm not sure how we would achieve that. I guess it would require enums and generics, and probably not be very beautiful. Maybe @djbusstop has an idea?

File: src/components/layout/organize/TabbedLayout.tsx [hunk 3]
```diff
@@ -73,19 +68,24 @@ const OrganizeTabbedLayout: FunctionComponent<OrganizeTabbedLayoutProps> = ({ ch
...
-                            { tabs.map(tab => (
-                                <Tab key={ tab } label={ intl.formatMessage({
-                                    id: `layout.organize.campaigns.${tab}`,
-                                }) } value={ tab }
-                                />
-                            )) }
+                            { tabs.map(tab => {
+                                return (
+                                    <Tab key={ tab.label } label={ intl.formatMessage({
+                                        id: `layout.organize.campaigns.${tab.label}`,
```
Comment (richardolsson) [inline src/components/layout/organize/TabbedLayout.tsx:84]: The prefix here is not ideal IMO. It requires that all tabs are below the `layout.organize.campaign` group of messages. Instead, choose one of two options:

1. Let `tab.label` be the human-readable text, and use `intl.formatMessage()` outside of this component to get the localized strings and pass them in.
2. Rename `tab.label` to `tab.messageId` or something like that, and use `intl.formatMessage({ id: tab.messageId }) here.

File: src/components/layout/organize/AllCampaignsLayout.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+const AllCampaignsLayout: FunctionComponent<AllCampaignsLayoutProps> = ({ children, fixedHeight }) => {
+    const { orgId } = useRouter().query;
+    const intl = useIntl();
+
+    return (
+        <TabbedLayout
+            baseHref={ `/organize/${orgId}/campaigns` }
+            fixedHeight={ fixedHeight }
+            subtitle="subtitle for all campaigns tabs" //TODO: generate subtitle
```
Comment (richardolsson) [inline src/components/layout/organize/AllCampaignsLayout.tsx:19]: I don't know if these pages need subtitles?

File: src/components/layout/organize/SingleCampaignLayout.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+const SingleCampaignLayout: FunctionComponent<SingleCampaignLayoutProps> = ({ children, fixedHeight }) => {
+    const { campId, orgId } = useRouter().query;
+    const campaignQuery = useQuery(['campaign', orgId, campId ], getCampaign(orgId  as string, campId as string));
+
+    return (
+        <TabbedLayout
+            baseHref={ `/organize/${orgId}/campaigns/${campId}` }
+            fixedHeight={ fixedHeight }
+            subtitle="subtitle for single campaign tabs" //TODO: generate subtitle
```
Comment (richardolsson) [inline src/components/layout/organize/SingleCampaignLayout.tsx:20]: Can this be easily done using the existing code for the date range on the summary page? Otherwise lets clear the string, but keep the `TODO` comment.

**Review by richardolsson — CHANGES_REQUESTED** [review]
I think structure/architecture wise this looks mostly good, but I would like to change some things.

File: src/components/layout/organize/AllCampaignsLayout.tsx [hunk 6]
```diff
@@ -15,12 +15,12 @@ const AllCampaignsLayout: FunctionComponent<AllCampaignsLayoutProps> = ({ childr
...
+            defaultTab="summary"
             fixedHeight={ fixedHeight }
-            subtitle="subtitle for all campaigns tabs" //TODO: generate subtitle
             tabs={ [
-                {  defaultTab: true, href: `/`, label: 'summary' },
-                { href: `/calendar`, label: 'calendar' },
-                { href: `/archive`, label: 'archive' },
+                { href: `/`, label: 'summary', messageId: 'layout.organize.campaigns.summary' },
+                { href: `/calendar`, label: 'calendar', messageId: 'layout.organize.campaigns.calendar' },
+                { href: `/archive`, label: 'archive', messageId: 'layout.organize.campaigns.archive' },
```
Comment (richardolsson) [inline src/components/layout/organize/AllCampaignsLayout.tsx:23]: I'm starting to reconsider this. Here are my objections, and the conclusion:

The `label` property used to be for deciding the (eventually) human-readable label of the tab. Now it's more like a key or ID, but still called label.

The only reason why we need it is so that we can set the `defaultTab`. But I feel like we could use the `href` for that, couldn't we?

In fact, if we use `href` and just set the default value for `defaultTab` to `'/'`, we would very rarely have to change it.

So in conclusion, I think we should remove `label` and use `href` as the "key" instead. Do you think that's a good idea?

**Review by richardolsson — CHANGES_REQUESTED** [review]
I'm very sorry but there's one more thing that came to me now that I saw the changes to the `tabs` property.

File: src/components/layout/organize/AllCampaignsLayout.tsx (hunk 6 above)
Comment (nehahirve) [reply src/components/layout/organize/AllCampaignsLayout.tsx:23]: no you're right, much better this way. all three properties are unique for each tab and only href and messageId are needed

**Review by richardolsson — APPROVED** [review]
All good! Merge at will. 🟢
